### ISSUE TITLE
Type hints for PyG Data

### ISSUE THREAD

**Issue by a-r-j** [pyg-team/pytorch_geometric#5172]
### 🚀 The feature, motivation and pitch

Hiya, thanks for maintaining such a great project.

I've often run into scenarios where I have multiple configurations of features used within a project. I think a 'typehinting' scheme (and possibly validation) would be a great addition. I don't think we necessarily need to check for types (I think it's a reasonable assumption that everything will be an appropriate tensor but this could certainly be something to consider) but it would certainly be useful to check whether or not a `Data` object contains the relevant tensors.

This has started making it's way into various DL frameworks ((torch)[https://github.com/patrick-kidger/torchtyping], (JAX)[https://github.com/google/jaxtyping])

It could behave something like:

```python
from torch_geometric.types import Data

def forward(x: Data[edge_index, node_feats, edge_feats]):
    print(x.node_feats)

# Enforce types
from torch_geometric.types import validate

@validate
def forward(x: Data[edge_index, node_feats, edge_feats]):
    print(x.node_feats)
# Raises TypeError if x does not contain x.edge_index, x.node_feats, x.edge_feats
```

I've previously put together something similar for NetworkX: https://github.com/a-r-j/graphtype and would be quite happy to help implement this if you think it would make a good addition.

**Comment by rusty1s** [comment]
Thanks for the issue. Really interesting. Besides defining the attributes of the `data`, it is also possible to define their types? There exists a lot of attributes that get added to a `data` object during batching - do we need to specify these as well?

Can you estimate the amount of work/code change that is needed to integrate?

**Comment by a-r-j** [comment]
Yes, it should be quite easy to define their types. The current impl in the project I linked allows for specification of a subset or the entirety.

E.g this could look like:

```python
@validate
def forward(x: Data[edge_index: torch.Tensor]): # X can only contain the edge_index tensor attr
    pass 

@validate
def forward(x: Data[edge_index: torch.Tensor, ...]): # X has to contain edge_index tensor amongst other attributes
    pass 
```

The solution I'd suggest for PyG would include type checking/hinting attributes considering what you've just said. Within that I think integration with torch typing to specify shapes would be the biggest impact for users (and readability downstream!).

E.g. 


```python
from torchtyping import TensorType

@validate
def forward(x: Data[features: TensorType["nodes" , "n_feats"] ...]): 
   pass
```

I think I could draft a basic PR over the next couple weeks. I'll look into including TorchTyping as well - I've not tried it out much so hard to estimate. Overall, it should be quite light and I should be capable of doing it with some support :)

**Comment by rusty1s** [comment]
Super! This sounds great. Please go ahead with it. I am happy to support you.

**Comment by rusty1s** [comment]
We can start without `TensorType` support for now if that makes things easier :)

**Comment by Padarn** [comment]
Wow this looks super cool! I'm really looking forward to seeing it.

**Comment by mananshah99** [comment]
This looks really cool, also happy to help take a look at PRs as it is implemented. Thank you :)
